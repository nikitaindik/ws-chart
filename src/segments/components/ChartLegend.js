import React from 'react';
import { timeFormat } from 'd3-time-format';
import cx from 'classnames';

import TimeIcon from './TimeIcon';
import style from './ChartLegend.module.css';

const LegendIconSquare = ({ color }) => (
  <div className={style.iconWrap} style={{ background: color }} />
);

const LegendIconLine = ({ color }) => (
  <div className={cx(style.iconWrap, style.iconWrapFlex)}>
    <div className={style.iconLine} style={{ background: color }} />
  </div>
);

const LegendItem = ({ icon, label, value, primary }) => (
  <div className={style.legendItem}>
    <div>{icon}</div>
    <div>{label}:</div>
    <div className={cx({ [style.primaryValue]: primary })}>{value}</div>
  </div>
);

const ChartLegend = ({ activeBarData, showLatestBarData, latestBarData }) => {
  let data;

  if (!showLatestBarData && activeBarData) {
    data = activeBarData;
  } else {
    data = latestBarData;
  }

  const { added, removed, segmentSize, timestamp } = data;

  const formatter = timeFormat('%b %d, %H:%M');
  const formattedTime = formatter(timestamp);

  return (
    <div className={style.legendWrap}>
      <div className={style.column}>
        <LegendItem
          icon={<LegendIconSquare color="#6c59d3" />}
          label={'Added'}
          value={added}
        />
        <LegendItem
          icon={<LegendIconSquare color="#999ea1" />}
          label={'Removed'}
          value={Math.abs(removed)}
        />
      </div>
      <div className={style.column}>
        <LegendItem
          icon={<LegendIconLine color="#008af7" />}
          label={'Size'}
          value={segmentSize}
          primary
        />
        {!showLatestBarData && (
          <LegendItem
            icon={<TimeIcon />}
            label={'Time'}
            value={formattedTime}
          />
        )}
      </div>
    </div>
  );
};

export default ChartLegend;
