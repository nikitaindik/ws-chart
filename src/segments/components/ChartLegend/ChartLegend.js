import React from 'react';
import { timeFormat } from 'd3-time-format';
import cx from 'classnames';

import { ReactComponent as TimeIcon } from '../icons/TimeIcon.svg';

import { colors } from '../../constants';

import style from './ChartLegend.module.css';

const formatter = timeFormat('%b %d, %H:%M');

const LegendIconSquare = ({ color }) => (
  <div className={style.iconWrap} style={{ background: color }} />
);

const LegendIconLine = ({ color }) => (
  <div className={cx(style.iconWrap, style.iconWrapFlex)}>
    <div className={style.iconLine} style={{ background: color }} />
  </div>
);

const LegendItem = ({ icon, label, value, primary }) => (
  <div className={style.legendItem}>
    <div>{icon}</div>
    <div>{label}:</div>
    <div className={cx({ [style.primaryValue]: primary })}>{value}</div>
  </div>
);

const ChartLegend = ({
  hoveredBarData,
  showLatestBarData,
  latestBarData,
  activeBarSize,
}) => {
  let data;

  if (!showLatestBarData && hoveredBarData) {
    data = hoveredBarData;
  } else {
    data = latestBarData;
  }

  const { added, removed, segmentSize } = data;
  const barStartTimestamp = data.timestamp;

  const formattedTime = showLatestBarData
    ? formatter(Date.now())
    : formatter(barStartTimestamp + activeBarSize);

  return (
    <div className={style.legendWrap}>
      <div className={style.column}>
        <LegendItem
          icon={<LegendIconSquare color={colors.CHART_POSITIVE_COLOR} />}
          label={'Added'}
          value={added}
        />
        <LegendItem
          icon={<LegendIconSquare color={colors.CHART_NEGATIVE_COLOR} />}
          label={'Removed'}
          value={Math.abs(removed)}
        />
      </div>
      <div className={style.column}>
        <LegendItem
          icon={<LegendIconLine color={colors.CHART_LINE_COLOR} />}
          label={'Size'}
          value={segmentSize}
          primary
        />
        <LegendItem icon={<TimeIcon />} label={'Time'} value={formattedTime} />
      </div>
    </div>
  );
};

export default ChartLegend;
