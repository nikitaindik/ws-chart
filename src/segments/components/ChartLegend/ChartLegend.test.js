import React from 'react';
import { render } from 'react-testing-library';

import ChartLegend from './ChartLegend';

test('ChartLegend snapshot', () => {
  const mockedProps = {
    hoveredBarData: {
      timestamp: 1,
      segmentId: '10',
      added: 1,
      removed: 1,
      segmentSize: 1,
    },
    latestBarData: {
      timestamp: 3,
      segmentId: '10',
      added: 1,
      removed: 1,
      segmentSize: 1,
    },
    showLatestBarData: false,
    activeBarSize: '360000',
  };

  const { container } = render(<ChartLegend {...mockedProps} />);
  expect(container).toMatchSnapshot();
});
