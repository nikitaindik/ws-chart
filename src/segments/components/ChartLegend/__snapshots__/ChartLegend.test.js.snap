// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChartLegend snapshot 1`] = `
<div>
  <div
    class="legendWrap"
  >
    <div
      class="column"
    >
      <div
        class="legendItem"
      >
        <div>
          <div
            class="iconWrap"
            style="background: rgb(108, 89, 211);"
          />
        </div>
        <div>
          Added
          :
        </div>
        <div
          class=""
        >
          1
        </div>
      </div>
      <div
        class="legendItem"
      >
        <div>
          <div
            class="iconWrap"
            style="background: rgb(153, 158, 161);"
          />
        </div>
        <div>
          Removed
          :
        </div>
        <div
          class=""
        >
          1
        </div>
      </div>
    </div>
    <div
      class="column"
    >
      <div
        class="legendItem"
      >
        <div>
          <div
            class="iconWrap iconWrapFlex"
          >
            <div
              class="iconLine"
              style="background: rgb(0, 138, 247);"
            />
          </div>
        </div>
        <div>
          Size
          :
        </div>
        <div
          class="primaryValue"
        >
          1
        </div>
      </div>
      <div
        class="legendItem"
      >
        <div>
          <svg>
            TimeIcon.svg
          </svg>
        </div>
        <div>
          Time
          :
        </div>
        <div
          class=""
        >
          Jan 01, 02:22
        </div>
      </div>
    </div>
  </div>
</div>
`;
